src/herder/Upgrades.cpp

```diff
@@ -734,32 +723,31 @@ prepareLiabilities(LedgerManager& ledgerManager, LedgerDelta& ld)
 }
 
 void
-Upgrades::applyVersionUpgrade(LedgerManager& ledgerManager, LedgerDelta& ld,
-                              uint32_t newVersion)
+Upgrades::applyVersionUpgrade(AbstractLedgerState& ls, uint32_t newVersion)
 {
-    LedgerHeader& header = ld.getHeader();
-    uint32_t prevVersion = header.ledgerVersion;
+    auto header = ls.loadHeader();
+    uint32_t prevVersion = header.current().ledgerVersion;
 
-    header.ledgerVersion = newVersion;
-    ledgerManager.getCurrentLedgerHeader().ledgerVersion = newVersion;
-    if (header.ledgerVersion >= 10 && prevVersion < 10)
+    header.current().ledgerVersion = newVersion;
+    if (header.current().ledgerVersion >= 10 && prevVersion < 10)
     {
-        prepareLiabilities(ledgerManager, ld);
+        header.deactivate();
```

Comment:
If I understand it correctly, header needs to be deactivated here only because it does not go out of scope here? If so, could we extract decision if prepareLiabilities should be called to a lambda, avoiding manual deactivation of header?

Like

```
auto shouldPrepare = [&](){
  auto header = ls.loadHeader();
  uint32_t prevVersion = header.current().ledgerVersion;
  header.current().ledgerVersion = newVersion;
  return header.current().ledgerVersion >= 10 && prevVersion < 10;
}();
if (shouldPrepare) { prepareLiabilities(ls); }
```

Not sure if that is better, but answer to that question will improve my confidence in how things works now.

Reply:
I think option 2. is OK, especially as prepareLiabilities is only called when we already have header instance.